### DISCUSSION TITLE
Feature - API caching

### DISCUSSION THREAD

**Discussion by meinte** [losol/eventuras#733] [General]
### The feature

Currently the frontend is not using any form of caching when retrieving information from the API - potentially causing more API calls than strictly necessary. Ideally any data that has already been fetched no longer requires a call to the API.

This feature came about after discussing the possibility of storing event registrations in user context. The mechanism for caching reservations however, needs to be a little bit smarter than simply caching the first results from the api. Consider this scenario:

- Customer visits his own profile page - reservations are fetched (and cached)
- Customer successfully makes a reservation
- Customer, upon visiting his profile page (or the event page) does not see any change - the reservations does not seem to be there
- The reason for that is the cached version has gotten 'dirty' and is no longer valid after creating the reservation

The following solutions describe the implementation from that reservations point-of-view, but would of course be reused in reducing other/future api calls.

### Ideal Solution

The ideal flow(the one that would introduce the minimum amount of api calls) would do the following:

- Reservations are fetched and cached
- A reservation is made - the return value (which should be the created reservation) is then added to the cached list of reservations
- Client side(cached) reservations are still up-to-date, without refetching the list of reservations

Downside - a complexer implementation which could potentially require fine-tuning per type of resource.

### Less ideal solution

- Reservations are fetched and cached
- Reservation is made
- The entire cache is invalidated and needs to be refetched

Downside - more API calls than strictly necessary compared to the ideal solution. Upside is that this solution is less complex.

### Other solutions
One thing that could be implemented, but would only cater to resources which are more static is caching headers. Vercel has some custom headers which can be set and could reduce round-trips to the eventuras server. This could be implemented in the api forwarder.

https://vercel.com/docs/edge-network/caching

**Comment by losolio** [comment]
Thanks for raising this issue! It is an important discussion, and will convert this to discussions!
